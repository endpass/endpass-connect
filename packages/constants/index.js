export const INPAGE_EVENTS = {
  SETTINGS: 'INPAGE_PROVIDER_SETTINGS_EVENT',
  RESPONSE: 'INPAGE_PROVIDER_RESPONSE_EVENT',
  REQUEST: 'INPAGE_PROVIDER_REQUEST_EVENT',
};
export const INPAGE_ID_PREFIX = 'ep_';
export const HOST_WINDOW_NAME = 'endpass-connect-host';
export const DIALOG_WINDOW_NAME = 'endpass-connect-dialog';
export const STORAGE_USER_META_KEY = 'endpass-user-meta';
export const AVAILABLE_USER_META_PROPS = ['activeAccount'];
export const DAPP_WHITELISTED_METHODS = [
  'personal_sign',
  'eth_personalSign',
  'eth_signTypedData',
  'eth_sendTransaction',
];
