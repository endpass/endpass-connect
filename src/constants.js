export const NET_ID = Object.freeze({
  MAIN: 1,
  ROPSTEN: 3,
  RINKEBY: 4,
  ETHEREUM_CLASSIC: 61,
});

/**
 * Networks map
 */
export const NETWORK_URL = Object.freeze({
  ETH: [
    // 'wss://eth-mainnet.endpass.com:2084',
    // 'wss://eth-mainnet.endpass.com',
    'https://eth-mainnet.endpass.com:2083',
    `https://mainnet.infura.io/${ENV.infura.key}`,
  ],
  ROP: [
    // 'wss://eth-ropsten.endpass.com:2084',
    // 'wss://eth-ropsten.endpass.com',
    'https://eth-ropsten.endpass.com:2083',
    `https://ropsten.infura.io/${ENV.infura.key}`,
  ],
  RIN: [`https://rinkeby.infura.io/${ENV.infura.key}`],
  ETC: [
    // 'wss://etc-mainnet.endpass.com:2084',
    // 'wss://etc-mainnet.endpass.com',
    'https://etc-mainnet.endpass.com:2083',
    'https://etc-geth.0xinfra.com',
  ],
});

export const DEFAULT_NETWORKS = Object.freeze({
  [NET_ID.MAIN]: {
    id: NET_ID.MAIN,
    networkType: 'main',
    currency: 1,
    name: 'Main',
    url: NETWORK_URL.ETH,
  },
  [NET_ID.ROPSTEN]: {
    id: NET_ID.ROPSTEN,
    name: 'Ropsten',
    networkType: 'ropsten',
    currency: 2,
    url: NETWORK_URL.ROP,
  },
  [NET_ID.RINKEBY]: {
    id: NET_ID.RINKEBY,
    name: 'Rinkeby',
    networkType: 'rinkeby',
    currency: 2,
    url: NETWORK_URL.RIN,
  },
  [NET_ID.ETHEREUM_CLASSIC]: {
    id: NET_ID.ETHEREUM_CLASSIC,
    name: 'Ethereum classic',
    networkType: 'ethClassic',
    currency: 3,
    url: NETWORK_URL.ETC,
  },
});

/**
 * Static things
 */
export const INPAGE_EVENTS = {
  SETTINGS: 'INPAGE_PROVIDER_SETTINGS_EVENT',
  RESPONSE: 'INPAGE_PROVIDER_RESPONSE_EVENT',
  REQUEST: 'INPAGE_PROVIDER_REQUEST_EVENT',
  DROP_PENDING: 'INPAGE_PROVIDER_REQUEST_DROP_PENDING',
  LOGIN: 'INPAGE_PROVIDER_REQUEST_LOGIN',
  LOGGED_IN: 'INPAGE_PROVIDER_REQUEST_LOGGED_IN',
};
export const INPAGE_ID_PREFIX = 'ep_';
export const DAPP_WHITELISTED_METHODS = [
  'eth_sign',
  'personal_sign',
  'personal_ecRecover',
  'eth_personalSign',
  'eth_signTypedData',
  'eth_sendTransaction',
];

export const METHODS = {
  AUTH: 'AUTH',
  SIGN: 'SIGN',
  LOGOUT: 'LOGOUT',
  ACCOUNT: 'ACCOUNT',
  RECOVER: 'RECOVER',
  GET_SETTINGS: 'GET_SETTINGS',
  READY_STATE_DIALOG: 'READY_STATE_DIALOG',
  READY_STATE_BRIDGE: 'READY_STATE_BRIDGE',
};
