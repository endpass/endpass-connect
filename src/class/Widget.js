import { WIDGET_EVENTS } from '@/constants';
import { inlineStyles } from '@/util/dom';

const FADE_TIMEOUT = 300;
const getStylesByPosition = position => {
  return (
    position || {
      bottom: '15px',
      right: '15px',
    }
  );
};
const createWidgetIframeStyles = position => {
  return inlineStyles(
    Object.assign(
      {
        position: 'fixed',
        'z-index': 6000000,
        width: '240px',
        height: '70px',
        border: 'none',
        'border-radius': '4px',
        opacity: 0,
        transition: `opacity ${FADE_TIMEOUT}ms ease-in`,
      },
      getStylesByPosition(position),
    ),
  );
};

export default class Widget {
  /**
   * @param {Context} options.context Context
   * @param {String} options.url Context
   */
  constructor({ context, url }) {
    this.context = context;
    this.url = url;
    this.frame = null;

    this.handleWidgetFrameLoad = this.handleWidgetFrameLoad.bind(this);
  }

  /**
   * Create markup and prepend to <body>
   */
  mount(parameters = {}) {
    const { url } = this;
    const widgetMessenger = this.context.getWidgetMessenger();
    const styles = createWidgetIframeStyles(parameters.position);
    const markup = `
      <iframe id="endpass-widget" data-endpass="widget-frame" style="${styles}" src="${url}"></iframe>
    `;

    document.body.insertAdjacentHTML('afterBegin', markup);

    this.frame = document.body.querySelector('[data-endpass="widget-frame"]');
    this.frame.addEventListener('load', this.handleWidgetFrameLoad);

    widgetMessenger.setTarget(this.frame.contentWindow);
  }

  unmount() {
    this.frame.style.opacity = 0;

    setTimeout(() => {
      this.emitFrameEvent(WIDGET_EVENTS.DESTROY);
      this.frame.removeEventListener('load', this.handleWidgetFrameLoad);
      this.frame.remove();
    }, FADE_TIMEOUT);
  }

  getWidget() {
    return new Promise(resolve => {
      /* eslint-disable-next-line */
      const handler = setTimeout(() => {
        if (this.frame) {
          return resolve(this.frame);
        }

        handler();
      }, 250);
    });
  }

  handleWidgetFrameLoad() {
    this.emitFrameEvent(WIDGET_EVENTS.MOUNT);
    this.frame.style.opacity = 1;
  }

  emitFrameEvent(event, detail) {
    const frameEvent = new CustomEvent(event, {
      detail,
    });

    this.frame.dispatchEvent(frameEvent);
  }

  /**
   * @param {String} options.height Height in px or other CSS/HTML friendly unit
   */
  resize({ height }) {
    this.frame.style.height = height;
  }
}
