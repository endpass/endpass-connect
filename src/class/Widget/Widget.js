import CrossWindowMessenger from '@endpass/class/CrossWindowMessenger';
import debounce from 'lodash.debounce';
import { DIRECTION, METHODS, WIDGET_EVENTS } from '@/constants';
import { inlineStyles } from '@/util/dom';
import {
  MOBILE_BREAKPOINT,
  FADE_TIMEOUT,
  INITIAL_FRAME_STYLES,
  FRAME_DESKTOP_STYLES,
  FRAME_MOBILE_COLLAPSED_STYLES,
  FRAME_MOBILE_EXPANDED_STYLES,
} from './WidgetStyles';

export default class Widget {
  /**
   * @param {object} options
   * @param {string} options.namespace namespace of connect
   * @param {string} options.url frame url
   */
  constructor({ namespace, url }) {
    this.url = url;
    this.frame = null;
    this.position = null;

    this.isMounted = false;
    this.isLoaded = false;
    this.isExpanded = false;

    this.handleWidgetFrameLoad = this.handleWidgetFrameLoad.bind(this);
    this.handleDocumentClick = this.handleDocumentClick.bind(this);
    this.handleScreenResize = this.handleScreenResize.bind(this);
    this.debouncedHandleScreenResize = debounce(this.handleScreenResize, 100);
    this.widgetMessenger = new CrossWindowMessenger({
      showLogs: !ENV.isProduction,
      name: `connect-bridge-widget[${namespace}]`,
      to: DIRECTION.AUTH,
      from: DIRECTION.CONNECT,
    });
    this.subscribe();
  }

  /* eslint-disable-next-line */
  get isMobile() {
    return window.innerWidth < MOBILE_BREAKPOINT;
  }

  getWidgetMessenger() {
    return this.widgetMessenger;
  }

  subscribe() {
    const { widgetMessenger } = this;

    widgetMessenger.subscribe(METHODS.WIDGET_INIT, (payload, req) => {
      req.answer({
        position: this.position,
        isMobile: this.isMobile,
      });
    });

    widgetMessenger.subscribe(METHODS.WIDGET_EXPAND_REQUEST, (payload, req) => {
      this.isExpanded = true;
      this.handleDocumentClickOnce();
      this.frame.style = this.getWidgetFrameInlineStyles();
      req.answer();
    });
    widgetMessenger.subscribe(METHODS.WIDGET_COLLAPSE_REQUEST, () => {
      this.isExpanded = false;
      this.frame.style = this.getWidgetFrameInlineStyles();
    });

    widgetMessenger.subscribe(METHODS.WIDGET_OPEN, ({ root }, req) => {
      this.resize({ height: '100vh' });

      if (root) this.emitFrameEvent(WIDGET_EVENTS.OPEN);

      req.answer();
    });
    widgetMessenger.subscribe(METHODS.WIDGET_CLOSE, () => {
      this.emitFrameEvent(WIDGET_EVENTS.CLOSE);
    });
    widgetMessenger.subscribe(METHODS.WIDGET_FIT, ({ height }) => {
      this.resize({ height: `${height}px` });
    });
  }

  handleDocumentClick() {
    document.body.removeEventListener('click', this.handleDocumentClick);
    this.widgetMessenger.send(METHODS.WIDGET_COLLAPSE_RESPONSE);
  }

  handleDocumentClickOnce() {
    document.body.addEventListener('click', this.handleDocumentClick);
  }

  isWidgetMounted() {
    return this.isMounted;
  }

  emitFrameEvent(event, detail) {
    if (!this.frame) {
      return;
    }

    const frameEvent = new CustomEvent(event, {
      detail,
    });

    this.frame.dispatchEvent(frameEvent);
  }

  /**
   * Create markup and prepend to <body>
   */
  mount(parameters = {}) {
    const { url } = this;

    this.position = parameters.position || null;
    this.isMounted = true;

    const styles = this.getWidgetFrameInlineStyles();
    const markup = `
      <iframe id="endpass-widget" data-test="widget-frame" data-endpass="widget-frame" style="${styles}" src="${url}"></iframe>
    `;

    document.body.insertAdjacentHTML('afterBegin', markup);

    this.frame = document.body.querySelector('[data-endpass="widget-frame"]');
    this.frame.addEventListener('load', this.handleWidgetFrameLoad);
    window.addEventListener('resize', this.debouncedHandleScreenResize);

    this.widgetMessenger.setTarget(this.frame.contentWindow);

    return this.frame;
  }

  unmount() {
    // const widgetMessenger = this.widgetMessenger;

    this.frame.style.opacity = 0;
    this.isLoaded = false;
    this.isMounted = false;

    // widgetMessenger.unsubscribe(METHODS.WIDGET_INIT);
    // widgetMessenger.unsubscribe(METHODS.WIDGET_EXPAND_REQUEST);
    // widgetMessenger.unsubscribe(METHODS.WIDGET_OPEN);
    // widgetMessenger.unsubscribe(METHODS.WIDGET_CLOSE);
    // widgetMessenger.unsubscribe(METHODS.WIDGET_FIT);

    this.frame.removeEventListener('load', this.handleWidgetFrameLoad);
    window.removeEventListener('resize', this.debouncedHandleScreenResize);
    this.debouncedHandleScreenResize.cancel();

    // Awaiting application animation ending
    setTimeout(() => {
      this.emitFrameEvent(WIDGET_EVENTS.DESTROY);
      this.frame.remove();
      this.frame = null;
    }, FADE_TIMEOUT);
  }

  handleWidgetFrameLoad() {
    this.emitFrameEvent(WIDGET_EVENTS.MOUNT);
    this.isLoaded = true;
    this.frame.style = this.getWidgetFrameInlineStyles();
  }

  handleScreenResize() {
    this.widgetMessenger.send(METHODS.WIDGET_CHANGE_MOBILE_MODE, {
      isMobile: this.isMobile,
    });
    this.frame.style = this.getWidgetFrameInlineStyles();
  }

  /**
   * @param {object} options
   * @param {string} options.height Height in px or other CSS/HTML friendly unit
   */
  resize({ height }) {
    if (height) {
      this.frame.style.height = height;
    }
  }

  getWidgetNode() {
    return new Promise(resolve => {
      const handler = cb => {
        if (this.frame) {
          cb(this.frame);
          return true;
        }
        setTimeout(() => {
          handler(cb);
        }, 250);
        return false;
      };

      handler(resolve);
    });
  }

  getWidgetFrameDesktopPositionStylesObject() {
    const { position = {} } = this;
    const actualPosition = {
      left: 'auto',
      right: 'auto',
      top: 'auto',
      bottom: 'auto',
      ...position,
    };

    if (actualPosition.left === 'auto' && actualPosition.right === 'auto') {
      Object.assign(actualPosition, {
        right: '15px',
      });
    }

    if (actualPosition.top === 'auto' && actualPosition.bottom === 'auto') {
      Object.assign(actualPosition, {
        bottom: '5px',
      });
    }

    return actualPosition;
  }

  /**
   * @returns {object}
   */
  getWidgetFrameStylesObject() {
    const { isMobile, isExpanded, isLoaded } = this;

    switch (true) {
      case isMobile && isLoaded && isExpanded:
        return {
          ...FRAME_MOBILE_EXPANDED_STYLES,
          opacity: Number(this.isLoaded),
        };
      case isMobile && isLoaded && !isExpanded:
        return {
          ...FRAME_MOBILE_COLLAPSED_STYLES,
          opacity: Number(this.isLoaded),
        };
      case isLoaded:
        return {
          ...FRAME_DESKTOP_STYLES,
          ...this.getWidgetFrameDesktopPositionStylesObject(),
          opacity: Number(this.isLoaded),
        };
      default:
        return INITIAL_FRAME_STYLES;
    }
  }

  getWidgetFrameInlineStyles() {
    const stylesObject = this.getWidgetFrameStylesObject();

    if (!this.frame) {
      return inlineStyles(stylesObject);
    }

    const { clientHeight } = this.frame;

    Object.assign(stylesObject, {
      height: `${clientHeight}px`,
    });

    return inlineStyles(stylesObject);
  }
}
