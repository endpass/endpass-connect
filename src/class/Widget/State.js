/* eslint-disable class-methods-use-this */
// @ts-check
// eslint-disable-next-line no-unused-vars
import Widget from './Widget';

export default class State {
  /**
   *
   * @param {Widget} widget
   */
  constructor(widget) {
    this.widget = widget;
  }

  onExpand() {}

  onCollapse() {}

  onClose() {}

  onOpen() {}
}
