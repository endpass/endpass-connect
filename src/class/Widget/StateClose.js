import State from './State';

export default class StateClose extends State {
  onOpen(root) {
    this.widget.onOpen(root);
  }
}
