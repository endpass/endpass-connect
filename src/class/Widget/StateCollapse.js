import State from './State';

export default class StateCollapse extends State {
  onExpand() {
    this.widget.onExpand();
  }
}
