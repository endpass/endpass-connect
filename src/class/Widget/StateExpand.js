import State from './State';

export default class StateExpand extends State {
  onCollapse() {
    this.widget.onCollapse();
  }
}
