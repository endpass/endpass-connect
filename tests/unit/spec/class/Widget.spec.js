import Widget from '@/class/Widget';
import { METHODS, WIDGET_EVENTS } from '@/constants';

describe('Widget class', () => {
  const url = 'https://auth.foo.bar/public/widget';
  let messenger;
  let context;
  let widget;

  beforeEach(() => {
    messenger = {
      setTarget: jest.fn(),
      subscribe: jest.fn(),
    };
    context = {
      getWidgetMessenger: () => messenger,
    };
    widget = new Widget({ context, url });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('mount', () => {
    beforeEach(() => {
      jest.spyOn(document.body, 'insertAdjacentHTML');
      widget.subscribe = jest.fn();
    });

    it('should mount iframe for widget', () => {
      widget.mount();

      expect(document.body.insertAdjacentHTML.mock.calls[0][0]).toBe(
        'afterBegin',
      );
      expect(
        document.body.insertAdjacentHTML.mock.calls[0][1],
      ).toMatchSnapshot();
      expect(widget.subscribe).toBeCalled();
    });

    it('should mount iframe with given parameters', () => {
      widget.mount({ position: { top: '15px', left: '15px' } });

      expect(
        document.body.insertAdjacentHTML.mock.calls[0][1],
      ).toMatchSnapshot();
    });
  });

  describe('unmount', () => {
    beforeEach(() => {
      widget.subscribe = jest.fn();
    });

    it('should unmount widget after it faded out', () => {
      jest.useFakeTimers();

      widget.mount({ position: { top: '15px', left: '15px' } });
      widget.unmount();

      widget.emitFrameEvent = jest.fn();
      widget.frame.removeEventListener = jest.fn();
      widget.frame.remove = jest.fn();

      jest.advanceTimersByTime(300);

      expect(widget.emitFrameEvent).toBeCalledWith('destroy');
      expect(widget.frame.removeEventListener).toBeCalledWith(
        'load',
        expect.any(Function),
      );
      expect(widget.frame.remove).toBeCalled();
    });
  });

  describe('handleWidgetFrameLoad', () => {
    it('should emit load event and show widget', () => {
      widget.mount();
      const handler = jest.fn();
      widget.frame.addEventListener(WIDGET_EVENTS.MOUNT, handler);

      widget.handleWidgetFrameLoad();

      expect(handler).toBeCalled();
      expect(widget.frame.style.opacity).toBe('1');
    });
  });

  describe('emitFrameEvent', () => {
    it('should emit frame event through frame element', () => {
      widget.mount();
      const handler = jest.fn();
      widget.frame.addEventListener('foo', handler);

      const handlerNotCall = jest.fn();
      widget.frame.addEventListener('bar', handlerNotCall);

      widget.emitFrameEvent('foo', {
        bar: 'baz',
      });

      expect(handler).toBeCalledWith(expect.any(Object));
      expect(handlerNotCall).not.toBeCalled();
    });
  });

  describe('resize', () => {
    it('should resize frame', () => {
      widget.frame = {
        style: {
          height: 0,
        },
      };
      widget.resize({
        height: 300,
      });

      expect(widget.frame.style.height).toBe(300);
    });
  });

  describe('subscribe', () => {
    it('should subscribe on messenger view-responsible methods', () => {
      const contentWindow = 'foo';

      widget.frame = {
        contentWindow,
      };
      widget.subscribe();

      expect(messenger.setTarget).toBeCalledWith(contentWindow);
      expect(messenger.subscribe).toBeCalledWith(
        METHODS.WIDGET_OPEN,
        expect.any(Function),
      );
      expect(messenger.subscribe).toBeCalledWith(
        METHODS.WIDGET_CLOSE,
        expect.any(Function),
      );
      expect(messenger.subscribe).toBeCalledWith(
        METHODS.WIDGET_FIT,
        expect.any(Function),
      );
    });
  });

  describe('getWidgetNode', () => {
    beforeAll(() => {
      jest.useFakeTimers();
    });

    it('should returns promise which resolves with widget element node', done => {
      expect.assertions(1);

      const frameNode = {
        foo: 'bar',
      };

      widget.frame = frameNode;
      widget.getWidgetNode().then(res => {
        expect(res).toEqual(frameNode);
        done();
      });

      jest.advanceTimersByTime(500);
    });
  });
});
