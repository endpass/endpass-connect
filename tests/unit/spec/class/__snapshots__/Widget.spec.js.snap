// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget class mount should mount iframe for widget 1`] = `
"
      <iframe id=\\"endpass-widget\\" data-endpass=\\"widget-frame\\" style=\\"position: fixed;z-index: 99999999999999;width: 240px;height: 70px;border: none;border-radius: 4px;opacity: 0;transition: opacity 300ms ease-in;bottom: 15px;right: 15px;\\" src=\\"https://auth.foo.bar/public/widget\\"></iframe>
    "
`;

exports[`Widget class mount should mount iframe with given parameters 1`] = `
"
      <iframe id=\\"endpass-widget\\" data-endpass=\\"widget-frame\\" style=\\"position: fixed;z-index: 99999999999999;width: 240px;height: 70px;border: none;border-radius: 4px;opacity: 0;transition: opacity 300ms ease-in;top: 15px;left: 15px;\\" src=\\"https://auth.foo.bar/public/widget\\"></iframe>
    "
`;
